api/ratelimit/v1alpha1/ratelimit_types.go

```diff
@@ -0,0 +1,149 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of HTTPRoute.
+	Spec RatelimitSpec `json:"spec"`
+
+	// Status defines the current state of Ratelimit.
+	Status RatelimitStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// RatelimitList contains a list of Ratelimit resources.
+type RatelimitList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Ratelimit `json:"items"`
+}
+
+// RatelimitSpec defines the desired state of Ratelimit
+type RatelimitSpec struct {
+	// Type decides the scope for the Ratelimits.
+	Type RatelimitType `json:"type"`
+	// Rules are a list of Ratelimit matchers and limits.
+	//
+	// +kubebuilder:validation:MaxItems=16
+	Rules []RatelimitRule `json:"rules"`
+}
+
+// RatelimitType specifies the types of Ratelimiting.
+// Valid RatelimitType values are:
+//
+// * "Local"
+// * "Global"
+//
+// +kubebuilder:validation:Enum=Local;Global
+type RatelimitType string
+
+const (
+	// In this mode, the ratelimits are applied per Envoy proxy instance.
+	RatelimitTypeLocal RatelimitType = "Local"
+
+	// In this mode, the ratelimits are applied across all Envoy proxy instances.
+	RatelimitTypeGlobal RatelimitType = "Global"
+)
+
+// RatelimitRule defines the semantics for matching attributes
+// from the incoming requests, and setting limits for them.
+type RatelimitRule struct {
+	// +optional
+	// +kubebuilder:validation:MaxItems=8
+	Matches []RatelimitMatch `json:"matches,omitempty"`
+	Limit   RatelimitValue   `json:"limit"`
+}
+
+// RatelimitMatch specifies the attributes within the traffic flow that can
+// be matched on.
+type RatelimitMatch struct {
+	// ClientAddress defines the semantics for matching on the IP Address
+	// of the client making the request.
+	// +optional
+	ClientAddress *ClientAddressMatch `json:"clientAddress,omitempty"`
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Headers []HeaderMatch `json:"headers,omitempty"`
+}
+
+// ClientAddressMatch defines the match attributes based on the source IP Address
+// originated from the client.
+type ClientAddressMatch struct {
+	// Cidr specifies the subnet range of source IP addresses
+	// to be matched on.
+	// Not setting this field, implies matching on all source/client IP addresses.
+	// +optional
+	Cidr *string `json:"cidr,omitempty"`
+}
+
+// HeaderMatch defines the match attributes within the HTTP Headers of the request.
+type HeaderMatch struct {
+	// Type specifies how to match against the value of the header.
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *HeaderMatchType `json:"type,omitempty"`
+
+	// Name of the HTTP header.
+	Name string `json:"name"`
+
+	// Value within the HTTP header.
+	// Not setting this field, implies matching on all unique values within the header.
+	// +optional
+	Value *string `json:"value,omitempty"`
```

Comment:
support `CaseSensitive`?

Reply:
good point, will add a note mentioning case insensitivity to match the Gateway API spec
https://github.com/kubernetes-sigs/gateway-api/blob/9568c673895dd3ff0c2b1d50029702af619c7f2c/apis/v1beta1/httproute_types.go#L377